### ISSUE TITLE
After receiving HTTP 500 for the first request, the client starts sending requests without an upgrade header

### ISSUE THREAD

**Issue by p-himik** [taoensso/sente#460]
My WS handler for GET:
```clj
(fn [req]
  (println "returning HTTP 500")
  (clojure.pprint/pprint (:headers req))
  (res/status 500))
```

Cleaned-up logged data:
```
returning HTTP 500
{"origin"                   "http://localhost:8000",
 "host"                     "localhost:8000",
 "upgrade"                  "websocket",
 "user-agent"               "...",
 "cookie"                   "...",
 "connection"               "Upgrade",
 "pragma"                   "no-cache",
 "sec-websocket-key"        "...",
 "accept-language"          "en-US,en;q=0.9",
 "sec-websocket-version"    "13",
 "accept-encoding"          "gzip, deflate, br, zstd",
 "sec-websocket-extensions" "permessage-deflate; client_max_window_bits",
 "cache-control"            "no-cache"}

returning HTTP 500
{"sec-fetch-site"     "same-origin",
 "x-requested-with"   "XMLHTTPRequest",
 "sec-ch-ua-mobile"   "?0",
 "host"               "localhost:8000",
 "user-agent"         "...",
 "cookie"             "...",
 "sec-ch-ua"          "...",
 "sec-ch-ua-platform" "\"Linux\"",
 "referer"            "...",
 "connection"         "keep-alive",
 "x-csrf-token"       "test-csrf-token",
 "accept"             "*/*",
 "accept-language"    "en-US,en;q=0.9",
 "sec-fetch-dest"     "empty",
 "accept-encoding"    "gzip, deflate, br, zstd",
 "sec-fetch-mode"     "cors"}

[all the following logged headers are identical]
```

In the initial description of the issue, I formulated it as "the client spams the server", but it was just a dev-time configuration in effect where the retry time is a small constant.

**Comment by iarenaza** [comment]
@p-himik This looks to me like the `:type` option was set to `:auto`, in the sente client side. Which means that it tries using the Websocket method first (as can be seen by the `Upgrade: websocket` header, and the `Sec-Websocket-*` headers). 
But if it cannot establish a successful websocket connection, it falls back to the AJAX connection method. Which seems to be the case, as the `X-Requested-With` headers says it's an XMLHTTPRequest, and the `Sec-Fetch-Mode` header is set to `cors` (which is an indication that this is an AJAX request, as Websocket connections don't use CORS at tall).

If you se `:type` to `:ws` in the sente client side, the client will retry connecting via websockets only, instead of falling back to AJAX.

**Comment by p-himik** [comment]
I've probably cleaned up the data a bit too eagerly.

AFAIU, if it were just switching over to AJAX, the POST handler would've been called.
But in this case, the very same GET handler is called.

**Comment by iarenaza** [comment]
Not really.  The initial handshake uses `GET` in both cases (websockets and AJAX[1]). Once the handshake is done, getting data from the server is done by "long polling" on the "-chsk-connect!" `GET`  requests sent by the AJAX client to the server.

As far as I can tell, the AJAX client only uses `POST` for sending data to the server[2]. 

Here is a screen capture from a system where we have configure the client with the `:type` setting set to `:auto`, where I (intentionally) force an error on the Websocket connection attempt, but accept the AJAX connection attempt (I did this back in the day to make sure that we handled the fallback correctly on both sides). This is taken from the Firefox developer console:

![Image](https://github.com/user-attachments/assets/d3f3b4af-cac1-4fdb-8c32-3fc95a641a33)

Notice how all of the connections (the initial one via websocket and the sub-sequent XHR ones, see "Initiator" column in the screenshot) use the `GET` method. And how the initial "handshake" connection (the one having the extra `&handshake=true` parameter) takes only 6 ms to complete. While the rest of the sub-sequent request take 20 seconds (the default "long polling" value, that we haven't changed).

[1] https://github.com/taoensso/sente/blob/v1.20.0/src/taoensso/sente.cljc#L1746
[2] https://github.com/taoensso/sente/blob/v1.20.0/src/taoensso/sente.cljc#L1683

**Comment by iarenaza** [comment]
Here's a similar screenshot, with the `type` setting changed to `:ws` (same system, same configuration, only that setting changed on the client side):

![Image](https://github.com/user-attachments/assets/a98c11e6-303a-492e-90c2-41aa54238b98)

Notice how in this case, all the connection attempts use the Websocket "Initiator".

**Comment by iarenaza** [comment]
And sure enough, all of those requests have the "Upgrade" header. Here I picked on at random, just to show it:

![Image](https://github.com/user-attachments/assets/5adcbe19-5e05-4899-a331-e5c8b4fa2974)

**Comment by p-himik** [comment]
Ah, I see. So I completely misunderstood how the AJAX part works then, and this misunderstanding has been with me for years.

I wonder why AJAX doesn't use POST for everything... On the surface, it seems that it would make things simpler.

In any case, it seems that this issue can be closed then.
